cabal-install/Distribution/Solver/Modular/Solver.hs

```diff
@@ -101,4 +124,87 @@ solve sc cinfo idx pkgConfigDB userPrefs userConstraints userGoals =
                                                   , PackageName "integer-simple"
                                                   ])
-    buildPhase       = addLinking $ buildTree idx (independentGoals sc) userGoals
+    buildPhase       = traceTree "build.json" id
+                     $ addLinking
+                     $ buildTree idx (independentGoals sc) userGoals
+
+-- | Dump solver tree to a file (in debugging mode)
+--
+-- This only does something if the @debug-tracetree@ configure argument was
+-- given; otherwise this is just the identity function.
+traceTree ::
+#ifdef DEBUG_TRACETREE
+  GSimpleTree a =>
+#endif
+     FilePath  -- ^ Output file
+  -> (a -> a)  -- ^ Function to summarize the tree before dumping
+  -> a -> a
+#ifdef DEBUG_TRACETREE
+traceTree = gtraceJson
+#else
+traceTree _ _ = id
+#endif
+
+#ifdef DEBUG_TRACETREE
+instance GSimpleTree (Tree QGoalReason) where
+  fromGeneric = go
+    where
+      go :: Tree QGoalReason -> SimpleTree
+      go (PChoice qpn _     psq) = Node "P" $ Assoc $ L.map (uncurry (goP qpn)) $ PSQ.toList psq
+      go (FChoice _   _ _ _ psq) = Node "F" $ Assoc $ L.map (uncurry goFS)      $ PSQ.toList psq
+      go (SChoice _   _ _   psq) = Node "S" $ Assoc $ L.map (uncurry goFS)      $ PSQ.toList psq
+      go (GoalChoice        psq) = Node "G" $ Assoc $ L.map (uncurry goG)       $ PSQ.toList psq
+      go (Done _rdm)             = Node "D" $ Assoc []
+      go (Fail cs _reason)       = Node "X" $ Assoc [("CS", Leaf $ goCS cs)]
+
+      -- Show package choice
+      goP :: QPN -> POption -> Tree QGoalReason -> (String, SimpleTree)
+      goP _        (POption (I ver _loc) Nothing)  subtree = (showVersion ver, go subtree)
+      goP (Q _ pn) (POption _           (Just pp)) subtree = (showQPN (Q pp pn), go subtree)
+
+      -- Show flag or stanza choice
+      goFS :: Bool -> Tree QGoalReason -> (String, SimpleTree)
+      goFS val subtree = (show val, go subtree)
+
+      -- Show goal choice
+      goG :: Goal QPN -> Tree QGoalReason -> (String, SimpleTree)
+      goG (Goal var gr) subtree = (showVar var ++ " (" ++ shortGR gr ++ ")", go subtree)
+
+      -- Variation on 'showGR' that produces shorter strings
+      -- (Actually, QGoalReason records more info than necessary: we only need
+      -- to know the variable that introduced the goal, not the value assigned
+      -- to that variable)
+      shortGR :: QGoalReason -> String
+      shortGR UserGoal               = "user"
+      shortGR (PDependency (PI nm _)) = showQPN nm
+      shortGR (FDependency nm _)      = showQFN nm
+      shortGR (SDependency nm)        = showQSN nm
+
+      -- Show conflict set
+      goCS :: ConflictSet QPN -> String
+      goCS cs = "{" ++ (intercalate "," . L.map showVar . CS.toList $ cs) ++ "}"
+#endif
+
+-- | Replace all goal reasons with a dummy goal reason in the tree
+--
+-- This is useful for debugging (when experimenting with the impact of GRs)
+_removeGR :: Tree QGoalReason -> Tree QGoalReason
```

Comment:
If this is not exported and not used anywhere, why are we not getting a Travis failure due to `-Werror`?